# Introduction 

<p style="text-align: justify"> In this jupyter notebook, the main objective is to explore the data from a glass formation dataset initially obtained by the <i> SciGlass Python</i> module, having been pre-processed by Daniel Cassar, and to make the appropriate changes for the use of machine learning methods on the dataset. This dataset initially contains more than 72k examples of glasses, crystals and intermediates compositions, in addition to 319 attributes associated with these materials. There are 317 quantitative attributes of inorganic compounds that may be present in the material, as well as 1 quantitative attribute of a unique code associated with each material (<code>Kod</code>) and 1 qualitative attribute about if the compounds form glass, crystals or intermediates (<code>GF</code>).

### Imported Required Librarie

In [1]:
import pandas as pd

## Exploratory Data Analytics

### Basic Exploratory Data Analytics and Basic Processing 

<p style="text-align: justify"> This subsection shows a brief analysis of the dataset, with a descriptive analysis, exclusion of duplicate examples and a better ordering of the data. Firstly, the file was imported from inside a zip folder and partially displayed to briefly observe it <sup>[1, 2, 3]</sup>.

In [2]:
df = pd.read_csv("../data/raw/GF_data_compound.zip", sep = "\t")
display(df)

,ZnO,Sb2O3,Na2O,MoO3,Bi2O3,TeO2,V2O5,P2O5,B2O3,CaO,...,BiNbO4,Se4I,GeTe4.3,GaTe3,ZnI2,Na2Se,MgS,AsTe3,Kod,GF
0,0.00,0.00,0.0,0.0,0.00,0.00,0.0,0.00,0.00,0.00,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,21883,Clear glass
1,0.00,0.00,0.0,0.0,0.00,88.55,0.0,0.00,0.00,0.00,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,21472,Clear glass
2,0.00,0.00,0.0,0.0,0.00,0.00,25.4,0.00,49.35,25.25,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,22665,Glass and crystal
3,0.00,0.00,0.0,0.0,89.70,0.00,0.0,0.00,0.00,0.00,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,23009,Glass and crystal
4,0.00,0.00,0.0,0.0,0.00,0.00,0.0,0.00,0.00,0.00,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,24639,Crystal
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
72121,0.00,49.96,0.0,0.0,0.00,29.78,0.0,0.00,20.26,0.00,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,22220,Crystal
72122,0.00,0.00,0.0,0.0,0.00,10.35,0.0,0.00,39.97,0.00,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,22603,Crystal
72123,20.28,0.00,0.0,0.0,0.00,0.00,0.0,30.02,0.00,0.00,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,33008,Glass
72124,0.00,0.00,0.0,0.0,48.89,0.00,0.0,0.00,0.00,0.00,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,22175,Clear glass



<p style="text-align: justify"> The quantitative attribute of a unique code for each example will be deleted because of the non-relevance in machine learning models. It will also check how many missing and negative elements there are, as well as how many duplicate examples or duplicate attributes there are. In addition, since the attributes are a percentage of the composition of the solid studied, it will be checked how many examples there are where the sum of the attributes doesn't result in 100% <sup>[3, 4, 5, 6]</sup>.

In [3]:
df = df.drop('Kod', axis=1) # drop the code attribute

print('Number of missing values:', df.isnull().sum().sum())
print('Number of negative values:', (((df.drop('GF', axis=1)) < 0).sum().sum()))

print('''Number of duplicated attributes:''', (df.columns.duplicated().sum()))
print('Number of duplicated examples:', df.duplicated().sum())
print('''Number of examples that the sum of attributes doesn't result in 100%:''', (round((df.drop('GF', axis=1)).sum(axis=1), 5) != 100).sum())

Number of missing values: 0
Number of negative values: 474
Number of duplicated attributes: 0
Number of duplicated examples: 1720
Number of examples that the sum of attributes doesn't result in 100%: 0


<p style="text-align: justify"> There are negative elements and duplicate examples, the duplicate examples are deleted. In addition, the number of examples with at least one negative value was checked and descriptive statistics of the dataset were displayed to better understand the negative values <sup>[3, 4, 5, 6]</sup>.


In [4]:
df = df.drop_duplicates(keep=False) # drop_duplicates

print('Number of examples with at least one negative value:', len(df[(df.select_dtypes(include='number') < 0).any(axis=1)]))

pd.set_option('display.max_rows', None)  # show all rows 
pd.set_option('display.max_columns', None) # show all columns

round(df.describe(), 2) # show descriptive statistics of the dataset

Number of examples with at least one negative value: 457


,ZnO,Sb2O3,Na2O,MoO3,Bi2O3,TeO2,V2O5,P2O5,B2O3,CaO,Nb2O5,Ge,Se,Ag,BaO,K2O,SrO,Li2O,La2O3,MgO,ThO2,Al2O3,BeO,ZrO2,TiO2,Ta2O5,WO3,Tl2O,CdO,PbO,In2O3,SnO2,Ga2O3,Cs2O,Rb2O,Y2O3,Sc2O3,SiO2,GeO2,UO2,LaF3,ZrF4,PbF2,Ag2O,MnO,Cu2O,HfO2,AlN,CrO3,Nd2O3,AlF3,Fe2O3,Cr2O3,BaF2,InF3,MgF2,PbCl2,ZnCl2,PbI2,Fe3O4,CeO2,CuO,ThF4,ZnF2,YbF3,YF3,As2O3,Dy2O3,NiO,CoO,LiCl,SrF2,CaF2,Sm2O3,BeF2,KF,VO2,NaF,HgO,MnO2,As2O5,SO3,Na2S,MnF2,KCl,ThCl4,NaCl,BiCl3,CdCl2,PbBr2,CuCl,BaCl2,CsI,BaI2,CuI,ZnBr2,KBr,CdBr2,BaBr2,MgBr2,CaBr2,SrBr2,MnBr2,CoBr2,FeBr2,BiF3,CsF,AgF,AgI,FeO,TmF3,NdF3,TlBr,TlI,GdF3,SnO,SnCl2,Sb,Mn2O3,DyF3,HoF3,ErF3,Sb2O5,Gd2O3,SnF2,Er2O3,Yb2O3,Co2O3,Ni2O3,NiSO4,CdF2,GaF3,UF4,Ca3N2,P3N5,Li3N,FeF3,FeF2,CdI2,KI,PrF3,SmF3,CeF3,EuF3,TbF3,AgBr,AgCl,CrF3,Li2SO4,NaI,SeO3,Pr2O3,S,Ga,As,Cu,EuS,Ga2S3,GeS2,As2Se3,GeTe,I,B,Ni,P,Tl,Bi,BaS,La2S3,Te,Al,Ag2S,GeS,Cl,ZnS,CdS,CdTe,Sb2Se3,GeSe2,Si,Cd,Br,MnS,ZnSe,GeSe3,Tl2S,Sb2S3,As2S3,Cu2Se,SnS,PbS,PbSe,GeSe,Li2S,In,MnCl2,HgTe,Hg,Sn,CaS,SrS,Fe,CdSe,Pb,B2S3,P2S5,SnSe,SnSe2,AsI3,SbI3,BiI3,Fe2S3O12,ZnSO4,K2SO4,AsSe,H2S,SbO2,Ag2Se,HgSe,O,LiI,SiS2,LiBr,Ag4SSe,SnTe,NaPO3,CuPO3,Cu2WO4,Cu3PO4,CuBr,Cu2MoO4,AgPO3,Ag2MoO4,HgI2,As2Te3,BaHPO4,BaP2O6,NbO2F,LiF,BPO4,Sb2Te3,TeI,In2S3,K2S,ZrS2,Ga2Se3,In2Se3,CsBr,Zn,Nd2S3,Eu,BaB2O4,BaGeO3,Al2S3,Na2SO4,PbP2O6,ZnP2O6,Na2B4O7,MgSO4,Bi2S3,AlP3O9,K2NbOF5,Li3PO4,Al2Y2O6,PbTe,CsCl,Li2MoO4,Li2WO4,Na2MoO4,Na2WO4,Ag2Te,Bi2Se3,BaPO3F,ZnTe,HgS,AlPO4,Ho2O3,Li2B4O7,Tl2Se,SrB2O4,Pr6O11,Tb4O7,ErCl3,GdCl3,TmCl3,HoCl3,NdCl3,KPO3,SrSO4,ErPO4,SbP3O9,ScP3O9,HfF4,KHSO4,KBO2,GeTe4,In2Te6,Ga2Te3,KAsO3,Na4As2O7,Co3O4,Er2S3,LiHSO4,NaHSO4,SrCl2,TaO2F,Lu2O3,AsTe,P2Se5,BiNbO4,Se4I,GeTe4.3,GaTe3,ZnI2,Na2Se,MgS,AsTe3
count,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.0,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.0,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.

<p style="text-align: justify"> It's important to understand that negative values are errors, it's impossible to have negative values as a composition. In this sense, the presence of these values is probably human error, these values should most likely be zero. Therefore, the negative numbers will be set to zero and the rows that had negative values will have all the values normalized so that the sum of the attributes results in 100%. After this, it will be checked if there are still negative values and if the sum of the attributes of any row does not result in 100% <sup>[3, 7, 8]</sup>.

In [5]:
categorical_column = df['GF'] # The GF column will be temporarily separated.

df = df.drop('GF', axis=1) # drop "GF" column 

df[df < 0] = 0 # negative numbers are equal to zero

df = df.div(df.sum(axis=1), axis=0) * 100 # normalize the values ​​of each row

print('Number of negative values:', ((df < 0).sum().sum()))

print('''Number of examples that the sum of attributes doesn't result in 100%:''', (round(df.sum(axis=1), 5) != 100).sum())

df = pd.concat([df, categorical_column], axis=1) # add the 'GF' column back to the dataset

Number of negative values: 0
Number of examples that the sum of attributes doesn't result in 100%: 0


<p style="text-align: justify"> Finally, the index will be reset and the size and descriptive statistics of the current dataset will be displayed <sup>[1, 2, 9]</sup>.

In [6]:
df = df.reset_index(drop=True)

print('Dimensionality of the DataFrame:', df.shape) # show the dimensionality of the DataFrame
round(df.describe(), 2)

Dimensionality of the DataFrame: (69092, 318)


,ZnO,Sb2O3,Na2O,MoO3,Bi2O3,TeO2,V2O5,P2O5,B2O3,CaO,Nb2O5,Ge,Se,Ag,BaO,K2O,SrO,Li2O,La2O3,MgO,ThO2,Al2O3,BeO,ZrO2,TiO2,Ta2O5,WO3,Tl2O,CdO,PbO,In2O3,SnO2,Ga2O3,Cs2O,Rb2O,Y2O3,Sc2O3,SiO2,GeO2,UO2,LaF3,ZrF4,PbF2,Ag2O,MnO,Cu2O,HfO2,AlN,CrO3,Nd2O3,AlF3,Fe2O3,Cr2O3,BaF2,InF3,MgF2,PbCl2,ZnCl2,PbI2,Fe3O4,CeO2,CuO,ThF4,ZnF2,YbF3,YF3,As2O3,Dy2O3,NiO,CoO,LiCl,SrF2,CaF2,Sm2O3,BeF2,KF,VO2,NaF,HgO,MnO2,As2O5,SO3,Na2S,MnF2,KCl,ThCl4,NaCl,BiCl3,CdCl2,PbBr2,CuCl,BaCl2,CsI,BaI2,CuI,ZnBr2,KBr,CdBr2,BaBr2,MgBr2,CaBr2,SrBr2,MnBr2,CoBr2,FeBr2,BiF3,CsF,AgF,AgI,FeO,TmF3,NdF3,TlBr,TlI,GdF3,SnO,SnCl2,Sb,Mn2O3,DyF3,HoF3,ErF3,Sb2O5,Gd2O3,SnF2,Er2O3,Yb2O3,Co2O3,Ni2O3,NiSO4,CdF2,GaF3,UF4,Ca3N2,P3N5,Li3N,FeF3,FeF2,CdI2,KI,PrF3,SmF3,CeF3,EuF3,TbF3,AgBr,AgCl,CrF3,Li2SO4,NaI,SeO3,Pr2O3,S,Ga,As,Cu,EuS,Ga2S3,GeS2,As2Se3,GeTe,I,B,Ni,P,Tl,Bi,BaS,La2S3,Te,Al,Ag2S,GeS,Cl,ZnS,CdS,CdTe,Sb2Se3,GeSe2,Si,Cd,Br,MnS,ZnSe,GeSe3,Tl2S,Sb2S3,As2S3,Cu2Se,SnS,PbS,PbSe,GeSe,Li2S,In,MnCl2,HgTe,Hg,Sn,CaS,SrS,Fe,CdSe,Pb,B2S3,P2S5,SnSe,SnSe2,AsI3,SbI3,BiI3,Fe2S3O12,ZnSO4,K2SO4,AsSe,H2S,SbO2,Ag2Se,HgSe,O,LiI,SiS2,LiBr,Ag4SSe,SnTe,NaPO3,CuPO3,Cu2WO4,Cu3PO4,CuBr,Cu2MoO4,AgPO3,Ag2MoO4,HgI2,As2Te3,BaHPO4,BaP2O6,NbO2F,LiF,BPO4,Sb2Te3,TeI,In2S3,K2S,ZrS2,Ga2Se3,In2Se3,CsBr,Zn,Nd2S3,Eu,BaB2O4,BaGeO3,Al2S3,Na2SO4,PbP2O6,ZnP2O6,Na2B4O7,MgSO4,Bi2S3,AlP3O9,K2NbOF5,Li3PO4,Al2Y2O6,PbTe,CsCl,Li2MoO4,Li2WO4,Na2MoO4,Na2WO4,Ag2Te,Bi2Se3,BaPO3F,ZnTe,HgS,AlPO4,Ho2O3,Li2B4O7,Tl2Se,SrB2O4,Pr6O11,Tb4O7,ErCl3,GdCl3,TmCl3,HoCl3,NdCl3,KPO3,SrSO4,ErPO4,SbP3O9,ScP3O9,HfF4,KHSO4,KBO2,GeTe4,In2Te6,Ga2Te3,KAsO3,Na4As2O7,Co3O4,Er2S3,LiHSO4,NaHSO4,SrCl2,TaO2F,Lu2O3,AsTe,P2Se5,BiNbO4,Se4I,GeTe4.3,GaTe3,ZnI2,Na2Se,MgS,AsTe3
count,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.0,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.0,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.00,69092.

## Analysis of Numeric Attributes and <code>GF</code> Labels, and Sorting the Dataset

### Basic Analysis of Numeric Attributes and <code>GF</code> Labels

<p style="text-align: justify"> All the numerical attributes of the dataset were show:

In [7]:
for column in df.columns:
    print(column)

ZnO
Sb2O3
Na2O
MoO3
Bi2O3
TeO2
V2O5
P2O5
B2O3
CaO
Nb2O5
Ge
Se
Ag
BaO
K2O
SrO
Li2O
La2O3
MgO
ThO2
Al2O3
BeO
ZrO2
TiO2
Ta2O5
WO3
Tl2O
CdO
PbO
In2O3
SnO2
Ga2O3
Cs2O
Rb2O
Y2O3
Sc2O3
SiO2
GeO2
UO2
LaF3
ZrF4
PbF2
Ag2O
MnO
Cu2O
HfO2
AlN
CrO3
Nd2O3
AlF3
Fe2O3
Cr2O3
BaF2
InF3
MgF2
PbCl2
ZnCl2
PbI2
Fe3O4
CeO2
CuO
ThF4
ZnF2
YbF3
YF3
As2O3
Dy2O3
NiO
CoO
LiCl
SrF2
CaF2
Sm2O3
BeF2
KF
VO2
NaF
HgO
MnO2
As2O5
SO3
Na2S
MnF2
KCl
ThCl4
NaCl
BiCl3
CdCl2
PbBr2
CuCl
BaCl2
CsI
BaI2
CuI
ZnBr2
KBr
CdBr2
BaBr2
MgBr2
CaBr2
SrBr2
MnBr2
CoBr2
FeBr2
BiF3
CsF
AgF
AgI
FeO
TmF3
NdF3
TlBr
TlI
GdF3
SnO
SnCl2
Sb
Mn2O3
DyF3
HoF3
ErF3
Sb2O5
Gd2O3
SnF2
Er2O3
Yb2O3
Co2O3
Ni2O3
NiSO4
CdF2
GaF3
UF4
Ca3N2
P3N5
Li3N
FeF3
FeF2
CdI2
KI
PrF3
SmF3
CeF3
EuF3
TbF3
AgBr
AgCl
CrF3
Li2SO4
NaI
SeO3
Pr2O3
S
Ga
As
Cu
EuS
Ga2S3
GeS2
As2Se3
GeTe
I
B
Ni
P
Tl
Bi
BaS
La2S3
Te
Al
Ag2S
GeS
Cl
ZnS
CdS
CdTe
Sb2Se3
GeSe2
Si
Cd
Br
MnS
ZnSe
GeSe3
Tl2S
Sb2S3
As2S3
Cu2Se
SnS
PbS
PbSe
GeSe
Li2S
In
MnCl2
HgTe
Hg
Sn
CaS
SrS
Fe
CdSe
Pb
B2S3
P2S5
SnSe
SnSe2
As

<p style="text-align: justify"> All the <code>GF</code> labels and the number of times they are present in the dataset have been displayed. There are 209 labels in total. <sup>[10]</sup>

In [8]:
display(df['GF'].value_counts())

pd.reset_option('display.max_rows')
pd.reset_option('display.max_columns')

GF
Glass                                              25149
Crystal                                            14719
Clear glass                                        10592
Opaque glass                                        2760
Glass and crystal                                   2426
Glass+crystal                                       2013
Imperfect melt                                      1693
Partially opaque glass                              1245
Phase-separated glass                               1046
Glass containing a crystalline part                  719
Glass + crystal                                      701
Unmelted                                             432
Clear glass with crystalline part                    295
Stable glass                                         233
Crystalline                                          217
Opal glass                                           202
Phase separated glass                                181
Phase-separation            

### Sorting the Dataset

<p style="text-align: justify"> Finally, the dataset was sorted based on the <code>GF</code> attribute, using as a criterion the number of times each label is present in the dataset in descending order. <sup>[10]</sup>

In [9]:
df = df.iloc[df.groupby('GF').GF.transform('size').mul(-1).argsort(kind='mergesort',)].reset_index(drop=True)
df = df.reset_index(drop=True)
display(df)

,ZnO,Sb2O3,Na2O,MoO3,Bi2O3,TeO2,V2O5,P2O5,B2O3,CaO,...,P2Se5,BiNbO4,Se4I,GeTe4.3,GaTe3,ZnI2,Na2Se,MgS,AsTe3,GF
0,0.0,0.0,0.00,0.0,0.0,0.00,76.84,4.6,0.00,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,Glass
1,0.0,0.0,0.00,0.0,0.0,0.00,0.00,0.0,0.00,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,Glass
2,0.0,0.0,0.00,0.0,0.0,0.00,0.00,0.0,34.65,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,Glass
3,0.0,0.0,16.16,0.0,0.0,83.76,0.00,0.0,0.00,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,Glass
4,0.0,0.0,0.00,0.0,0.0,0.00,0.00,0.0,0.00,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,Glass
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
69087,0.0,0.0,0.00,0.0,0.0,0.00,0.00,0.0,0.00,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,Partially opacue glass
69088,0.0,0.0,0.00,0.0,0.0,0.00,0.00,0.0,0.00,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,Glass with bubbles
69089,0.0,0.0,0.00,0.0,0.0,0.00,0.00,0.0,0.00,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,Imprefect melt
69090,0.0,0.0,0.00,0.0,0.0,0.00,0.00,0.0,0.00,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,Glass containing a cryatalline part


## Selecting GF Labels

<p style="text-align: justify"> In the previous topic, you can see that there are 209 <code>GF</code> labels about what the composition forms: glass, crystal or an intermediate. To simplify things in a way that is probably more useful for applying machine learning models, the labels will be reduced to just three: crystal (<code>value -1</code>), glass (<code>value 1</code>) and intermediates (<code>value 0</code>). To do this, labels that appear at least 100 times will be analyzed and either classified into one of the three labels mentioned or just deleted. <sup>[11]</sup>

In [10]:
glass = ['Glass','Clear glass', 'Opaque glass', 'Partially opaque glass', 'Phase-separated glass', 'Stable glass', 'Opal glass', 'Phase separated glass', 
        'Phase-separation', 'Homogeneous glass', 'Heterogeneous glass', 'Opaque']
crystal = ['Crystal', 'Crystalline', 'Crystallization']
glass_and_crystal = ['Glass and crystal', 'Glass+crystal', 'Glass containing a crystalline part', 'Glass + crystal', 'Clear glass with crystalline part', 'Glass with crystal']
delete = ['Imperfect melt', 'Unmelted', 'Batch stone', 'Crystal or imperfect melt', 'Surface crystallization', 'Fully devitrified glass', 'Partial divitrification or opal glasses']

labels  = glass + crystal + glass_and_crystal

df = df[df['GF'].isin(labels)] 
df = df.replace(glass, 1).replace(crystal, -1).replace(glass_and_crystal, 0)

Check how many times the labels appear.<sup>[10]</sup>

In [11]:
display(df['GF'].value_counts())

GF
 1    41967
-1    15057
 0     6273
Name: count, dtype: int64

<p style="text-align: justify"> The dimensionality and the descriptive statistics of the current dataset will be displayed <sup>[1, 2, 9]</sup>.

In [12]:
print('Dimensionality of the DataFrame:', df.shape) # show the dimensionality of the DataFrame
round(df.describe(), 2)

Dimensionality of the DataFrame: (63297, 318)


,ZnO,Sb2O3,Na2O,MoO3,Bi2O3,TeO2,V2O5,P2O5,B2O3,CaO,...,P2Se5,BiNbO4,Se4I,GeTe4.3,GaTe3,ZnI2,Na2Se,MgS,AsTe3,GF
count,63297.00,63297.00,63297.00,63297.00,63297.00,63297.00,63297.00,63297.00,63297.00,63297.00,...,63297.00,63297.00,63297.00,63297.00,63297.00,63297.00,63297.00,63297.00,63297.00,63297.00
mean,1.86,0.71,2.22,0.61,1.85,9.21,2.43,4.14,10.63,1.71,...,0.01,0.02,0.01,0.01,0.00,0.00,0.00,0.01,0.01,0.43
std,8.70,6.17,9.04,5.18,8.92,24.26,11.42,13.69,21.99,8.17,...,0.59,1.04,0.51,1.02,0.30,0.14,0.31,0.60,0.57,0.85
min,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,...,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,-1.00
25%,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,...,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00
50%,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,...,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00
75%,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,...,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00
max,93.70,100.00,99.90,94.91,99.78,100.00,99.94,99.60,100.00,90.41,...,51.20,71.20,69.20,94.93,39.98,19.40,39.70,70.10,49.80,1.00


Finally, the dataset was saved as a <code>.csv</code> file.

In [13]:
df.to_csv("../data/processed/GFdata_compound_processed.csv", index=False)

## References

[1] Cassar, D. R. (2023). Aprendizagem de Máquina, 2.4 - Módulo pandas.

[2] Pereira, D. J. A., & CARVALHO, D. P. L. (2023). Aprendizagem de Máquina, Dados 03 - O Dataset Escolhido.

[3] khedekar, S. (2023, August 26). EDA on Netflix dataset using Pyhton. Kaggle. Retrieved February, 2024, from https://www.kaggle.com/code/swatikhedekar/eda-on-netflix-dataset-using-pyhton.

[4] Inada, I. (2023, August 12).  Counting the number of negative values in Pandas DataFrame. Retrieved February, 2024, from https://www.skytowner.com/explore/counting_the_number_of_negative_values_in_pandas_dataframe.

[5] Hunter, D. (2023b, April 15). Python: Filtering Negative Values in a DataFrame using Python Pandas. CopyProgramming. Retrieved February, 2024, from https://copyprogramming.com/howto/python-pandas-dataframe-filter-negative-values.

[6] note.nkmk.me (2024, January 26). pandas: Find, count, drop duplicates (duplicated, drop_duplicates). Retrieved February, 2024, from https://note.nkmk.me/en/python-pandas-duplicated-drop-duplicates/.

[7] Sharma, P. (2022, September 20) Python - Normalize rows of pandas dataframe by their sums. Retrieved March 2024, from https://www.includehelp.com/python/normalize-rows-of-pandas-dataframe-by-their-sums.aspx.

[8] FavTutor. (2021, November 17). Add Column to DataFrame Pandas (with Examples). Retrieved March 2024, from https://favtutor.com/blogs/pandas-add-column

[9] Saturn Cloud Blog. (2023b, December 12). How to reset index in a Pandas dataframe. Retrieved March 2024, from https://saturncloud.io/blog/how-to-reset-index-in-a-pandas-dataframe/

[10] Hall, P. (2023, May 23). Python: Sorting a Pandas data frame based on value counts in a column: a guide. CopyProgramming. Retrieved February 2024, from https://copyprogramming.com/howto/how-to-sort-a-pandas-data-frame-by-value-counts-of-a-column.

[11] Nelamali, N. (2023, November 29). Pandas filter rows by conditions. Spark by {Examples}. Retrieved March 2024, from https://sparkbyexamples.com/pandas/pandas-filter-rows-by-conditions/.